fabtests/unit/ep_test.c

```diff
@@ -0,0 +1,444 @@
+/*
+ * Copyright (c) 2019 Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <string.h>
+#include <time.h>
+
+#include <rdma/fi_errno.h>
+
+#include "unit_common.h"
+#include "shared.h"
+
+#define EP_TEST_SIZES_STRIDE 101
+
+static unsigned int num_eps, max_eps;
+struct fid_ep **ep_arr;
+struct fid_av **av_arr;
+struct fid_cq **txcq_arr, **rxcq_arr;
+static unsigned int seed = -1;
+
+static char err_buf[512];
+
+static int create_av(struct fid_av **av, enum fi_av_type type, int rx_ctx_bits,
+		     size_t count, size_t ep_per_node, const char *name,
+		     void *map_addr, uint64_t flags)
+{
+	struct fi_av_attr av_attr;
+
+	memset(&cq_attr, 0, sizeof(av_attr));
+	av_attr.type = type;
+	av_attr.rx_ctx_bits = rx_ctx_bits;
+	av_attr.count = count;
+	av_attr.ep_per_node = ep_per_node;
+	av_attr.name = name;
+	av_attr.map_addr = map_addr;
+	av_attr.flags = flags;
+
+	return fi_av_open(domain, &av_attr, av, NULL);
+}
+
+static int create_cq(struct fid_cq **cq, size_t size, uint64_t flags,
+		     enum fi_cq_format format, enum fi_wait_obj wait_obj)
+{
+	struct fi_cq_attr cq_attr;
+
+	memset(&cq_attr, 0, sizeof(cq_attr));
+	cq_attr.size = size;
+	cq_attr.flags = flags;
+	cq_attr.format = format;
+	cq_attr.wait_obj = wait_obj;
+
+	return fi_cq_open(domain, &cq_attr, cq, NULL);
+}
```

Comment:
Can ft_alloc_ep_res() be used instead of the above 2 calls?  That handles additional test options.  The call can be modified to take the av, cq, cntrs, etc. as input.